jmx/src/main/java/io/airlift/jmx/JmxAgent9.java

```diff
@@ -43,27 +44,71 @@
     public JmxAgent9(JmxConfig config)
             throws IOException
     {
+        Integer existingRegistryPort = Integer.getInteger(JMX_REGISTRY_PORT);
+        Integer existingServerPort = Integer.getInteger(JMX_SERVER_PORT);
+
+        if ((existingRegistryPort != null) && (config.getRmiRegistryPort() != null) &&
+                (!existingRegistryPort.equals(config.getRmiRegistryPort()))) {
+            throw new RuntimeException(format("System property [%s] value [%s] does match configured RMI registry port [%s]",
+                    JMX_REGISTRY_PORT, existingRegistryPort, config.getRmiRegistryPort()));
+        }
+
+        if ((existingServerPort != null) && (config.getRmiServerPort() != null) &&
+                (!existingServerPort.equals(config.getRmiServerPort()))) {
+            throw new RuntimeException(format("System property [%s] value [%s] does match configured RMI server port [%s]",
+                    JMX_SERVER_PORT, existingServerPort, config.getRmiServerPort()));
+        }
+
         int registryPort;
-        if (config.getRmiRegistryPort() == null) {
+        if (existingRegistryPort != null) {
+            registryPort = existingRegistryPort;
+        }
+        else if (config.getRmiRegistryPort() == null) {
             registryPort = NetUtils.findUnusedPort();
         }
         else {
             registryPort = config.getRmiRegistryPort();
         }
 
         int serverPort = 0;
-        if (config.getRmiServerPort() != null) {
+        if (existingServerPort != null) {
+            serverPort = existingServerPort;
+        }
+        else if (config.getRmiServerPort() != null) {
             serverPort = config.getRmiServerPort();
         }
 
+        // this is how the JDK JMX agent constructs its URL
+        JMXServiceURL jmxUrl = new JMXServiceURL("rmi", null, registryPort);
```

Comment:
This can fail for environmental reasons:
```java
if (host == null) {
    InetAddress local;
    try {
        local = InetAddress.getLocalHost();
    } catch (UnknownHostException e) {
        throw new MalformedURLException("Local host name unknown: " + e);
    }
    ...